src/modm/driver/encoder/bitbang_encoder_input.hpp

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) 2021, Thomas Sommer
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#ifndef MODM_BITBANG_ENCODER_INPUT_HPP
+#define MODM_BITBANG_ENCODER_INPUT_HPP
+
+#include <bit>
+#include <modm/platform.hpp>
+#include <type_traits>
+
+namespace modm
+{
+/**
+ * @ingroup modm_driver_bitbang_encoder_input
+ * @author  Thomas Sommer
+ *
+ * @brief					This driver decodes a AB (incremental) encoder signal
+ *
+ * @tparam SignalA          First modm::platform::Gpio pin to input the encoder signal.
+ * @tparam SignalB          Second modm::platform::Gpio pin to input the encoder signal.
+ * @tparam POSTSCALER	 	n_cycles to count as one in-/decrement.
+ * @tparam DeltaType	Must be signed integer and fit at least POSTSCALER. The Bigger
+ * 							DeltaType, the more inc-/decrements can be stored temporarily.
+ */
+template<typename SignalA, typename SignalB, uint8_t POSTSCALER = 4,
+		 std::signed_integral DeltaType = int8_t>
+class BitBangEncoderInput
+{
+	static_assert(std::popcount(POSTSCALER) == 1,
+				  "POSTSCALER must be an integer to basis 2 and not 0: 1, 2, 4, 8, 16, ...");
+	static_assert(POSTSCALER <= std::numeric_limits<DeltaType>::max(),
+				  "DeltaType is to small for POSTSCALER.");
+
+	using Phases = modm::platform::SoftwareGpioPort<SignalA, SignalB>;
+
```

Comment:
```suggestion
	static_assert(Phases::number_of_ports() == 1, "Signal A/B must be on the same GPIO port to prevent signal tearing!");
```
Otherwise the atomic read properties may not be true.

Reply:
Very nice mechanism.. Damn i ♥ love ♥ the way you've implemented the low level stuff

Reply:
I actually think it's just `Phases::number_of_ports` without a function call… I don't even remember the code I wrote.

Reply:
> Damn i ♥ love ♥ the way you've implemented the low level stuff

Thanks, half of the credit goes to C++17 with the relaxed constexpr rules:
https://github.com/modm-io/modm/pull/19#issuecomment-390767555

Reply:
> I actually think it's just `Phases::number_of_ports` without a function call… I don't even remember the code I wrote.

It is and has already been concidered in the last commits ;)

Reply:
> Thanks, half of the credit goes to C++17 with the relaxed constexpr rules:
Thanks for the interresting background!

Reply:
What's the magic spell to [get these shift maps](https://github.com/modm-io/modm/pull/19#issuecomment-390767555)?

Reply:
Here's the full example: https://github.com/modm-io/modm/blob/develop/examples/stm32f469_discovery/ports/main.cpp